### PR TITLE
DA-592 restrict non confirmed biobank samples

### PR DESCRIPTION
**Pull request by Seeker1911**

### CODE REVIEW

File: rest-api/dao/participant_summary_dao.py [hunk 1]
```diff
@@ -235,6 +235,7 @@ def update_from_biobank_stored_samples(self, participant_id=None):
       samples_to_isolate_dna = (
           CASE WHEN EXISTS(SELECT * FROM biobank_stored_sample
                            WHERE biobank_stored_sample.biobank_id = participant_summary.biobank_id
+                           AND biobank_stored_sample.confirmed IS NOT NULL
```
Comment (danqrodney) [inline rest-api/dao/participant_summary_dao.py]: This is only going to affect samples_to_isolate_dna. You also need to update the SQL for num_baseline_samples_arrived and in _get_sample_sql_and_params in here, as well as the reconciliation report:

https://github.com/all-of-us/raw-data-repository/blob/master/rest-api/offline/biobank_samples_pipeline.py#L302

Basically grep for biobank_stored_sample and make sure that all the places we reference it are filtering out samples without confirmed dates.

**Comment by danqrodney** [comment]
Other places to consider adding tests of confirmed date being null being treated the same as not having a sample at all:

https://github.com/all-of-us/raw-data-repository/blob/b37f97ec0cc7602f0f1f0daf816282841ee806af/rest-api/test/unit_test/offline_test/public_metrics_export_test.py#L130

https://github.com/all-of-us/raw-data-repository/blob/1f4c44487bc3c08c438dc3c9c1eb32588a8a3000/rest-api/test/unit_test/dao_test/participant_summary_dao_test.py#L226

https://github.com/all-of-us/raw-data-repository/blob/5062735a83ba48908dbe79e51bcb0e0e2de2bcce/rest-api/test/unit_test/offline_test/metrics_export_test.py#L177

https://github.com/all-of-us/raw-data-repository/blob/ec0a0ecb21ad331ce9334fcc0d4216319c52a009/rest-api/test/unit_test/api_test/participant_summary_api_test.py#L815

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py [hunk 2]
```diff
@@ -390,7 +410,7 @@ def test_reconciliation_query(self):
 
     # orders/samples where something went wrong; don't include orders/samples from more than 7
     # days ago, or where 24 hours hasn't elapsed yet.
-    exporter.assertRowCount(missing, 5)
+    exporter.assertRowCount(missing, 7)
```
Comment (danqrodney) [inline rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: Assert specifically the contents of the missing rows below.

I would have thought this would go up by 3, for each of the finalized tests in p_unconfirmed_samples_3.

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py [hunk 3]
```diff
@@ -166,9 +166,29 @@ def test_reconciliation_query(self):
...
+    # On time, recent order and samples not confirmed shows up in missing;
+    p_unconfirmed_samples = self._insert_participant()
+    self._insert_order(p_unconfirmed_samples, 'not_confirmed_order', BIOBANK_TESTS[:4], order_time,
+                       finalized_tests=BIOBANK_TESTS[:3], kit_id='kit3', tracking_number='t3',
+                       collected_note=u'\u2013foo', processed_note='bar', finalized_note='baz')
+    self._insert_samples(p_unconfirmed_samples, BIOBANK_TESTS[:3], ['Unconfirmed_sample',
+                         'Unconfirmed_sample2'], 'Ounconfirmed_sample', within_24_hours, None)
+
+    # old order time and samples not confirmed does not exist;
+    p_unconfirmed_samples_3 = self._insert_participant()
```
Comment (danqrodney) [inline rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: Try adding a non-confirmed sample without a matching order; this shouldn't show up in any reports.

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py [hunk 4]
```diff
@@ -166,9 +166,29 @@ def test_reconciliation_query(self):
     self._insert_samples(p_on_time, BIOBANK_TESTS[:2], ['GoodSample1', 'GoodSample2'], 'OGoodOrder',
                          within_24_hours, within_24_hours - datetime.timedelta(hours=1))
 
+
+
+    # On time, recent order and samples not confirmed shows up in missing;
+    p_unconfirmed_samples = self._insert_participant()
+    self._insert_order(p_unconfirmed_samples, 'not_confirmed_order', BIOBANK_TESTS[:4], order_time,
```
Comment (danqrodney) [inline rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: Do these actually show up in missing? It hasn't been 24 hours.... so I'm not sure they should.

File: rest-api/test/unit_test/offline_test/public_metrics_export_test.py [hunk 5]
```diff
@@ -150,6 +150,14 @@ def _create_data(self):
               test='1SAL',
               confirmed=TIME))
 
+      sample_dao.insert(
```
Comment (danqrodney) [inline rest-api/test/unit_test/offline_test/public_metrics_export_test.py]: So in order for this to be useful, we'd need to have a participant that had everything for being a full participant except 1SAL, and confirm that this stored sample (with no confirmed date) doesn't make them a full participant.

@calbach I can't see in the code where we're doing detailed tests of the public metrics stuff... do we have some somewhere?

**Comment by danqrodney** [comment]
I realized you need to add a check for confirmed_time IS NOT NULL here:

https://github.com/all-of-us/raw-data-repository/blob/d2c2b88c1b151f8eefda7ea9cc784a54b38e1e89/rest-api/offline/biobank_samples_pipeline.py#L438

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py (hunk 2 above)
Comment (danqrodney) [reply rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: Sorry, I realized that p_unconfirmed_samples_3 is inserted at order_time, too... so I would have thought that *none* of these would show up in missing, as it hasn't been 36 hours yet.

File: rest-api/dao/participant_summary_dao.py (hunk 1 above)
Comment (Seeker1911) [reply rest-api/dao/participant_summary_dao.py]: done

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py (hunk 4 above)
Comment (Seeker1911) [reply rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: corrected error in insert, good catch

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py (hunk 3 above)
Comment (Seeker1911) [reply rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: done

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py (hunk 2 above)
Comment (Seeker1911) [reply rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: correct, fixed insert.

File: rest-api/test/unit_test/offline_test/public_metrics_export_test.py (hunk 5 above)
Comment (Seeker1911) [reply rest-api/test/unit_test/offline_test/public_metrics_export_test.py]: havent done this

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py [hunk 6]
```diff
@@ -166,9 +166,31 @@ def test_reconciliation_query(self):
     self._insert_samples(p_on_time, BIOBANK_TESTS[:2], ['GoodSample1', 'GoodSample2'], 'OGoodOrder',
                          within_24_hours, within_24_hours - datetime.timedelta(hours=1))
 
+    # On time, recent order and samples not confirmed shows up in missing;
```
Comment (danqrodney) [inline rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: So these don't show up in missing, nor do the ones below. Please update the comment; and then add some orders which *do* show up in missing, because they use two_days_ago in place of order_time.

And then also add an unconfirmed sample which doesn't have a matching order, with confirmed=two_days_ago, and make sure it *doesn't* show up in missing.

Comment (Seeker1911) [reply rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: done, and as discussed...
confirmed (2 days ago) without order = missing

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py [hunk 7]
```diff
@@ -166,9 +166,46 @@ def test_reconciliation_query(self):
...
 
+    # On time, recent order and samples not confirmed do not show up in reports.
+    p_unconfirmed_samples = self._insert_participant()
+    self._insert_order(p_unconfirmed_samples, 'not_confirmed_order', BIOBANK_TESTS[:4], order_time,
+                       finalized_tests=BIOBANK_TESTS[:3], kit_id='kit3', tracking_number='t3',
+                       collected_note=u'\u2013foo', processed_note='bar', finalized_note='baz')
+    self._insert_samples(p_unconfirmed_samples, BIOBANK_TESTS[:3], ['Unconfirmed_sample',
+                         'Unconfirmed_sample2'], 'Ounconfirmed_sample', None, within_24_hours)
+
+    # two day old order not confirmed shows up in missing.
```
Comment (danqrodney) [inline rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: You should assert that the tests from this order specifically show up below.

File: rest-api/test/unit_test/offline_test/public_metrics_export_test.py (hunk 5 above)
Comment (danqrodney) [reply rest-api/test/unit_test/offline_test/public_metrics_export_test.py]: You can remove this for now. No sense in adding the sample without a matching test.

**Comment by danqrodney** [comment]
(We should do the metrics tests in a follow-up at some point.)

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py (hunk 7 above)
Comment (Seeker1911) [reply rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: done

File: rest-api/test/unit_test/offline_test/public_metrics_export_test.py (hunk 5 above)
Comment (Seeker1911) [reply rest-api/test/unit_test/offline_test/public_metrics_export_test.py]: done, merging.
